### PR TITLE
[2.10] Fix omapi_host on python3

### LINKED ISSUE #787: Fix omapi_host on python3
##### SUMMARY
omapi_host needs to use byte strings on python3 otherwise getting "TypeError: a bytes-like object is required, not 'str'" errors

Migrated from [#67559](https://github.com/ansible/ansible/issues/67559)

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
omapi_host module

##### ANSIBLE VERSION

```paste below
# ansible --version
ansible 2.9.4
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.6/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.6.8 (default, Oct 11 2019, 15:04:54) [GCC 8.3.1 20190507 (Red Hat 8.3.1-4)]
```
##### STEPS TO REPRODUCE
Create or remote a DHCP reservation using omapi_host module

##### EXPECTED RESULTS

Completed without errors

##### ACTUAL RESULTS

```paste below
"module_stdout": "Traceback (most recent call last):\r
  File \"/root/.ansible/tmp/ansible-tmp-1582083593.0190487-137096835301943/AnsiballZ_omapi_host.py\", line 102, in <module>\r
    _ansiballz_main()\r
  File \"/root/.ansible/tmp/ansible-tmp-1582083593.0190487-137096835301943/AnsiballZ_omapi_host.py\", line 94, in _ansiballz_main\r
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r
  File \"/root/.ansible/tmp/ansible-tmp-1582083593.0190487-137096835301943/AnsiballZ_omapi_host.py\", line 40, in invoke_module\r
    runpy.run_module(mod_name='ansible.modules.net_tools.omapi_host', init_globals=None, run_name='__main__', alter_sys=True)\r
  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\r
    return _run_module_code(code, init_globals, run_name, mod_spec)\r
  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\r
    mod_name, mod_spec, pkg_name, script_name)\r
  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\r
    exec(code, run_globals)\r
  File \"/tmp/ansible_omapi_host_payload_4n9lnp9o/ansible_omapi_host_payload.zip/ansible/modules/net_tools/omapi_host.py\", line 315, in <module>\r
  File \"/tmp/ansible_omapi_host_payload_4n9lnp9o/ansible_omapi_host_payload.zip/ansible/modules/net_tools/omapi_host.py\", line 304, in main\r
  File \"/tmp/ansible_omapi_host_payload_4n9lnp9o/ansible_omapi_host_payload.zip/ansible/modules/net_tools/omapi_host.py\", line 154, in __init__\r
  File \"/tmp/ansible_omapi_host_payload_4n9lnp9o/ansible_omapi_host_payload.zip/ansible/modules/net_tools/omapi_host.py\", line 159, in connect\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 980, in __init__\r
    self.initialize_authenticator(newauth)\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 1064, in initialize_authenticator\r
    response = self.query_server(msg)\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 1053, in query_server\r
    self.send_message(message)\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 1044, in send_message\r
    self.protocol.send_message(message, sign)\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 944, in send_message\r
    message.sign(self.authenticators[self.defauth])\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 466, in sign\r
    self.signature = authenticator.sign(self.as_string(forsigning=True))\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 457, in as_string\r
    self.serialize(ret, forsigning)\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 443, in serialize\r
    outbuffer.add_bindict(self.obj)\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 212, in add_bindict\r
    self.add_net16string(key).add_net32string(value)\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 183, in add_net32string\r
    return self.add_net32int(len(string)).add(string)\r
  File \"/root/.local/lib/python3.6/site-packages/pypureomapi.py\", line 145, in add\r
    self.buff.write(data)\r
TypeError: a bytes-like object is required, not 'str'\r"

```

### PR DESCRIPTION
**Pull request by amasolov**
##### SUMMARY

Fixes #787 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

module omapi_host

### CODE REVIEW

**Comment by amasolov** [comment]
ready_for_review

**Comment by felixfontein** [comment]
Can you add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)?

**Comment by amasolov** [comment]
@felixfontein done

File: changelogs/fragments/788-fix_omapi_host_on_python3.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - omapi_host - fix compatibility with python 3
```
Comment (felixfontein) [inline changelogs/fragments/788-fix_omapi_host_on_python3.yaml:2]:
```suggestion
  - omapi_host - fix compatibility with Python 3 (https://github.com/ansible-collections/community.general/pull/788).
```

**Comment by felixfontein** [comment]
Changelog entry looks good. I cannot really comment on the change. @nerzhul since it's your module, can you take a look?

**Comment by amasolov** [comment]
@felixfontein [same change](https://github.com/ansible/ansible/pull/67560) was previously approved by @nerzhul but didn't get merged for whatever reason

**Comment by felixfontein** [comment]
This PR has some differences from the one in ansible/ansible, most notable about using `to_bytes()` for dictionary keys. Is this actually necessary?

**Comment by amasolov** [comment]
@felixfontein yes, those are needed, didn't work otherwise. Most likely more work on python3 compatibility is needed in this module but I guess there is no much demand.

**Comment by Andersson007** [comment]
@amasolov let's get rid of merge commits

**Comment by amasolov** [comment]
@Andersson007 yup, done. Fair call, this one fell off the radar

**Comment by Andersson007** [comment]
@amasolov there's author's email in the header. If i were you, i'd email the author. Would be cool if anybody who uses this takes a look

**Comment by Riccardo1976** [comment]
Applied patch on ansible version 2.10.3,
Tested it and it works.

```
- name: add fake host
  omapi_host:
    key_name: omapikey
    key: "{{ omapi_host_key }}"
    host: 192.168.33.11
    macaddr: "00:50:56:cc:dd:ee"
    name: fake-host1
    ip: "192.168.33.15"
    state: present
    statements:
    - filename "pxelinux.0"
    - next-server 1.2.3.4
  delegate_to: localhost
```

output (with changing status:
```
host fake-host1 {
  dynamic;
  hardware ethernet 00:50:56:cc:dd:ee;
  fixed-address 192.168.33.15;
        supersede server.filename = "pxelinux.0";
}
host fake-host1 {
  dynamic;
  deleted;
}
host fake-host1 {
  dynamic;
  hardware ethernet 00:50:56:cc:dd:ee;
  fixed-address 192.168.33.15;
        supersede server.filename = "pxelinux.0";
        supersede server.next-server = 01:02:03:04;
}
```

**Comment by felixfontein** [comment]
Restarting CI.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
Let's get this merged finally :)

**Comment by felixfontein** [comment]
@amasolov thanks for fixing this!
@Riccardo1976 @Andersson007 thanks a lot for testing and reviewing!

**Comment by Andersson007** [comment]
@amasolov would you like to be added as a maintainer for `omapi_host` to `.github/BOTMETA.yml`?
Your `shipit`/`LGTM` will be counted by the bot to merge fixes / new feature automatically (needs 2 from maintainers) and you'll get notifications about related events (issues/ PRs), nothing obligatory of course and i don't think there will be a lot of them. What do you think?

**Comment by amasolov** [comment]
@Andersson007 happy to do that. I use omapi quite a bit in my workflows, so I wouldn't mind taking care about it.

**Pull request merged**